### DISCUSSION TITLE
Valtio SSR brainstorming

### DISCUSSION THREAD

**Discussion by italodeandra** [pmndrs/valtio#154] [General]
Hi team, I'm currently brainstorming about the usage of valtio in NextJS (SSR). All the solutions I can find to have a reference goes around using the Context API.

I don't like using Context with valtio. For me it loses one of my main purposes of using valtio in the first place. I'm currently keeping valtio on client side only, but I really would like to change that.

I'm still doing some research and testing myself, I'll add it here when I find something. But does anyone have any insight on this? Anything we could discuss.

**Comment by italodeandra** [comment]
Okay. I recently asked myself if maybe someone would have mentioned SSR on this repo, but I'm still the only one.

For future reference, if anyone comes around this asking yourself how the f are you supposed to sync Valtio's SSR state, this is my method when using cookies:


Entire `theme.state.ts` file:
```ts
import { isBrowser } from "@italodeandra/ui/utils/isBrowser";
import pick from "lodash/pick";
import { setCookie } from "nookies";
import { proxy, snapshot, subscribe } from "valtio";

const themeState = proxy({
  isDark: false,
  toggle(isDark?: boolean) {
    themeState.isDark =
      typeof isDark === "boolean" ? isDark : !themeState.isDark;
  },
  hydrated: false,
});

subscribe(themeState, () => {
  setCookie(
    null,
    "theme",
    JSON.stringify(pick(snapshot(themeState), ["isDark"])),
    {
      path: "/",
      maxAge: 60 * 60 * 24 * 365,
    }
  );
});

export const hydrateThemeState = (cookies?: { theme?: string }): void => {
  if (!themeState.hydrated && cookies?.theme) {
    themeState.hydrated = isBrowser;
    Object.assign(themeState, JSON.parse(cookies.theme));
  }
};
```

Fragment of `_app.tsx` file:

```tsx
import themeState, { hydrateThemeState } from "../src/theme.state";

const App: VFC<AppProps> = ({
  Component,
  pageProps,
}) => {
  hydrateThemeState(pageProps.cookies);

  const { isDark } = useSnapshot(themeState);

  return (
    <ThemeProvider theme={isDark ? darkTheme : lightTheme}>
      <Component {...pageProps} />
    </ThemeProvider>
  )
```


So the main idea behind this is to hydrate the state using the data from `pageProps`. It will hydrate on both Server and Browser, but you should stop hydrating it when it reaches the browser (hence the line `themeState.hydrated = isBrowser;`) so your state doesn't get stuck with old data.

**Comment by mfellner** [comment]
What would you say are the unique challenges that valtio is facing with SSR? Is there a feature that other state management libraries have and that valtio is missing which would be helpful in this respect?

One challenge I found in the context of a React Native app was to initialise proxy state from an initial value (e.g. locally persisted state or in your case server-side state).

The factory pattern implemented in [valtio-factory](https://github.com/mfellner/valtio-factory#provide-initial-state-on-initialization) (discussed here: https://github.com/pmndrs/valtio/discussions/348) allows for a clear separation of store (state) declaration and initialisation.

For example, you can declare your stores separately and run two different initialisation code paths for the server and the client side.

Since your question seems to be specific to Next.js, do you think it may be useful to add an example to their repository (https://github.com/vercel/next.js/tree/canary/examples)?

**Reply by italodeandra** [reply]
> What would you say are the unique challenges that valtio is facing with SSR? Is there a feature that other state management libraries have and that valtio is missing which would be helpful in this respect?

Valtio is currently not missing anything for my use cases. My main goal of this discussion was to try to use what Valtio has to offer to do things differently than the other state management libraries.

All state management libraries, no exception, needs Context API when facing SSR. But with this implementation I commented above I was able to avoid using Context completely.

> One challenge I found in the context of a React Native app was to initialise proxy state from an initial value (e.g. locally persisted state or in your case server-side state).
>
> The factory pattern implemented in [valtio-factory](https://github.com/mfellner/valtio-factory#provide-initial-state-on-initialization) (discussed here: https://github.com/pmndrs/valtio/discussions/348) allows for a clear separation of store (state) declaration and initialisation.

Your factory implementation seems really nice and well-thought-out, but on my use cases I just export a function like this

```ts
export const createStore = () => {
  const state = proxy({
    isDark: false,
    toggle(isDark?: boolean) {
      state.isDark =
        typeof isDark === "boolean" ? isDark : !state.isDark;
    },
  })
  subscribe(state, () => {
    // do something
  })
  return state
}
```

So I failed to see why it would be useful for me.

> For example, you can declare your stores separately and run two different initialisation code paths for the server and the client side.

But why would I have a separate store for server and client? The main issue is to hydrate the client state with data from the server.

> Since your question seems to be specific to Next.js, do you think it may be useful to add an example to their repository (https://github.com/vercel/next.js/tree/canary/examples)?

That's a nice idea, but I don't see my solution as the best we can offer because I created it by myself. That's why I created this discussion, to think of the best way together, but no one seems interested.

**Reply by Noitidart** [reply]
Can you please show how we have to do this with context api? I didn't know that all state libraries needed to use context for ssr.

**Reply by dai-shi** [reply]
FYI: There's a long thread in zustand, reaching finally to providing `zustand/context`. 

Unlike zustand, valtio doesn't need such a hack to use context: https://github.com/pmndrs/valtio/wiki/How-to-use-with-context

**Reply by italodeandra** [reply]
> Can you please show how we have to do this with context api?

Daishi sent a really good example above, please take a look.

But I still prefer not using the Context API, as I don't seem to actually need it if you try my example. Feels way more "valtio-y".

**Reply by Noitidart** [reply]
Ah thanks both! But why is context needed specifically for SSR, I'm slowly getting into SSR right now.

**Reply by dai-shi** [reply]
> FYI: There's a long thread in zustand, reaching finally to providing `zustand/context`.

Oops, forgot to attach the link: https://github.com/pmndrs/zustand/issues/182

**Reply by Noitidart** [reply]
Oh thanks that has all the content I need!

**Reply by wiesson** [reply]
What is the preferred way from `getServerSideProps` to a valtio state if I use nextjs?

**Reply by italodeandra** [reply]
> What is the preferred way from `getServerSideProps` to a valtio state if I use nextjs?

Hi @wiesson, the idea of this discussion was to decide an answer for this very question. There is no preferred or best way. But here you can see my approach for it: https://github.com/pmndrs/valtio/discussions/154#discussioncomment-2177435
